library/psa_crypto.c

```diff
@@ -1000,44 +1031,66 @@ psa_status_t psa_import_key_into_slot( psa_key_slot_t *slot,
             return( status );
 
         /* Allocate memory for the key */
-        status = psa_allocate_buffer_to_slot( slot, data_length );
+        status = psa_copy_key_material_into_slot( slot, data, data_length );
         if( status != PSA_SUCCESS )
             return( status );
 
-        /* copy key into allocated buffer */
-        memcpy( slot->data.key.data, data, data_length );
-
         /* Write the actual key size to the slot.
          * psa_start_key_creation() wrote the size declared by the
          * caller, which may be 0 (meaning unspecified) or wrong. */
         slot->attr.bits = (psa_key_bits_t) bit_size;
     }
-    else if( PSA_KEY_TYPE_IS_ECC( slot->attr.type ) )
-    {
+    else if( PSA_KEY_TYPE_IS_ASYMMETRIC( slot->attr.type ) )
+    {
+        /* Try validation through accelerators first. */
+        bit_size = slot->attr.bits;
+        psa_key_attributes_t attributes = {
+          .core = slot->attr
+        };
+        status = psa_driver_wrapper_validate_key( &attributes,
+                                                  data,
+                                                  data_length,
+                                                  &bit_size );
+        if( status == PSA_SUCCESS )
+        {
+            /* Key has been validated successfully by an accelerator.
+             * Copy key material into slot. */
+            status = psa_copy_key_material_into_slot( slot, data, data_length );
+            if( status != PSA_SUCCESS )
+                return( status );
+
+            slot->attr.bits = (psa_key_bits_t) bit_size;
+            return( PSA_SUCCESS );
+        }
+        else if( status != PSA_ERROR_NOT_SUPPORTED )
+            return( status );
+
+        /* Key format is not supported by any accelerator, try software fallback
+         * if present. */
+        if( PSA_KEY_TYPE_IS_ECC( slot->attr.type ) )
```

Comment:
Please move the `if( … )` condition inside the corresponding `#if defined(MBEDTLS_…)`. The code is correct either way, but this way is fragile (it depends on this if-chain running with `status` already set to `PSA_ERROR_NOT_SUPPORTED`) and wastes a few bytes of code (if you remove e.g. RSA support, there's no need to keep a call to `PSA_KEY_TYPE_IS_RSA`).

Reply:
I don't agree with it wasting code bytes (no statement in the conditional is volatile or in a different CU, so if the clause body is empty the compiler will optimise out the check as well), but if this really throws you off I can change it.

Reply:
@gilles-peskine-arm Force-pushed with your requested change.